plugins/fileservices/fileservices.cpp

```diff
@@ -2747,21 +2751,60 @@ static void checkExternalFileRights(ICodeContext *ctx, CDfsLogicalFileName &lfn,
     }
 }
 
-FILESERVICES_API void  FILESERVICES_CALL fsMoveExternalFile(ICodeContext * ctx,const char *location,const char *frompath,const char *topath)
+static void checkExternalFilePath(ICodeContext *ctx,IPropertyTree *plane,const char *host,
+    const char *path,bool checkScopePermissionIfDir,bool rd,bool wr,RemoteFilename &rfn)
 {
-    SocketEndpoint ep(location);
+    if (containsRelPaths(path)) //Detect a path like: a/../../../f
+        throw makeStringExceptionV(-1,"Invalid file path %s",path);
+
+    CDfsLogicalFileName dlfn;
+    if (plane)
+    {
+        //allow relative path if a plane name is specified.
+        if (isAbsolutePath(path))
+        {
+            const char *relativePath = getRelativePath(path,plane->queryProp("@prefix"));
+            if (nullptr == relativePath)
+                throw makeStringExceptionV(-1,"Invalid plane path %s.",path);
+            path = relativePath;
+        }
+        dlfn.setPlaneExternal(plane->queryProp("@name"),path);
+    }
+    else
+        dlfn.setExternal(host,path); //for backward compatibility
+    dlfn.getExternalFilename(rfn);
+
+    if (checkScopePermissionIfDir)
+    {
+        Owned<IFile> f = createIFile(rfn);
+        checkScopePermissionIfDir = f->isDirectory()==fileBool::foundYes;
+    }
+    checkExternalFileRights(ctx,dlfn,checkScopePermissionIfDir,rd,wr);
+}
+
+static IPropertyTree *checkPlaneOrHost(const char *planeName,const char *host)
+{
+    if (!isEmptyString(planeName))
+    {
+        Owned<IPropertyTree> plane = getDropZonePlane(planeName); //Only support DropZone for now.
+        if (!plane)
+            throw makeStringExceptionV(-1,"DropZone %s not found.",planeName);
+        return plane.getClear();
+    }
+    SocketEndpoint ep(host);
     if (ep.isNull())
-        throw MakeStringException(-1,"fsMoveExternalFile: Cannot resolve location %s",location);
-    CDfsLogicalFileName from;
-    from.setExternal(location,frompath);
-    CDfsLogicalFileName to;
-    to.setExternal(location,topath);
-    checkExternalFileRights(ctx,from,true,true);
-    checkExternalFileRights(ctx,to,false,true);
-    RemoteFilename fromrfn;
-    fromrfn.setPath(ep,frompath);
-    RemoteFilename torfn;
-    torfn.setPath(ep,topath);
+        throw makeStringExceptionV(-1,"Cannot resolve host %s",host);
+    return nullptr;
+}
+
+static void implementMoveExternalFile(ICodeContext *ctx,const char *location,
+    const char *frompath,const char *topath,const char *planename)
+{
+    Owned<IPropertyTree> plane = checkPlaneOrHost(planename,location);
+    RemoteFilename fromrfn,torfn;
+    checkExternalFilePath(ctx,plane,location,frompath,true,true,true,fromrfn);
+    checkExternalFilePath(ctx,plane,location,topath,false,false,true,torfn);
```

Comment:
see other comments, but if frompath is a directory, it implies topath should be too right?

Reply:
Yes. And the access to the directory (not a parent) should be checked.